<a href="https://colab.research.google.com/github/chayakonchanlun/DWDM21/blob/main/MiniExam.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### ข้อสอบ

จาก data ```flights.csv``` และ ```airports.csv``` ใน https://drive.google.com/drive/folders/1EFo1_uQgWkRiXjezL0T4YwteYrsqU5MV?usp=sharing
0. หารัฐที่มีจำนวนเครื่องบินขาเข้ามากที่สุดและน้อยที่สุด
1. หารัฐที่มีจำนวนเครื่องบินขาออกมากที่สุดและน้อยที่สุด
2. วาด Histogram แสดงจำนวนเครื่องบินขาออกของแต่ละวันในสัปดาห์
3. สายการบินไหนมีเที่ยวบินมากที่สุดและน้อยที่สุด
4. สายการบินอะไรมีความล่าช้ามากที่สุดและน้อยที่สุด
5. ช่วงเช้าหรือช่วงบ่ายมีเครื่องบินขาเข้ามากกว่ากัน

ให้นักศึกษาทำข้อสอบ 2 ข้อจากที่แสดงข้างบนโดยกรอกรหัสนักศึกษาลงใน function ```gen_exams()``` แล้ว run เพื่อให้ได้ข้อสอบที่นักศึกษาต้องทำ 2 ข้อ (ข้อละ 3 คะแนน)
และให้นักศึกษาคิดโจทย์เองอีก 1 ข้อ (4 คะแนน)


In [38]:
import pandas as pd

In [39]:
from google.colab import drive 
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [40]:
import os

In [41]:
flights = pd.read_csv('/content/drive/MyDrive/DWDM21_DATA/flights.csv')
flights

วันในเดือน  วันในสัปดาห์  ... ความล่าช้าขาออก  ความล่าช้าขาเข้า
0                19             5  ...              -3                 1
1                19             5  ...               0                -8
2                19             5  ...              -4               -15
3                19             5  ...              28                24
4                19             5  ...              -6               -11
...             ...           ...  ...             ...               ...
2702213          19             6  ...               1                -3
2702214          19             6  ...              -2                -3
2702215          19             6  ...              18                 9
2702216          19             6  ...              10                 7
2702217          19             6  ...               0               -13

[2702218 rows x 7 columns]

In [42]:
airports = pd.read_csv('/content/drive/MyDrive/DWDM21_DATA/airports.csv')
airports

รหัสสนามบิน         เมือง รัฐ                          ชื่อสนามบิน
0          10165   Adak Island  AK                                 Adak
1          10299     Anchorage  AK  Ted Stevens Anchorage International
2          10304         Aniak  AK                        Aniak Airport
3          10754        Barrow  AK      Wiley Post/Will Rogers Memorial
4          10551        Bethel  AK                       Bethel Airport
..           ...           ...  ..                                  ...
360        11233      Cheyenne  WY  Cheyenne Regional/Jerry Olson Field
361        11097          Cody  WY                 Yellowstone Regional
362        11865      Gillette  WY             Gillette Campbell County
363        12441       Jackson  WY                         Jackson Hole
364        14543  Rock Springs  WY       Rock Springs Sweetwater County

[365 rows x 4 columns]

In [43]:
def gen_exams(ID):
    import math
    a = int(ID[-1])%6
    b = int(ID[-1])%3
    print(f'รหัสนักศึกษา {ID} ทำข้อ {a} และ {b if a != b else 5-b}')

ตัวอย่างการ run

In [44]:
gen_exams('623020517-9')

รหัสนักศึกษา 623020517-9 ทำข้อ 3 และ 0


## **ข้อ 0 หารัฐที่มีจำนวนเครื่องบินขาเข้ามากที่สุดและน้อยที่สุด**

In [65]:
merged_table = airports.merge(flights,how='left',left_on='รหัสสนามบิน',right_on='รหัสสนามบินต้นทาง')
merged_table #(เชื่อมตารางโดยใช้ index ของ รหัสสนามบินเเละรหัสสนามบินต้นทาง)

รหัสสนามบิน         เมือง  ... ความล่าช้าขาออก ความล่าช้าขาเข้า
0              10165   Adak Island  ...             NaN              NaN
1              10299     Anchorage  ...            27.0              2.0
2              10299     Anchorage  ...           -10.0            -38.0
3              10299     Anchorage  ...            -3.0            -27.0
4              10299     Anchorage  ...            -6.0            -23.0
...              ...           ...  ...             ...              ...
2702508        11233      Cheyenne  ...             NaN              NaN
2702509        11097          Cody  ...             NaN              NaN
2702510        11865      Gillette  ...             NaN              NaN
2702511        12441       Jackson  ...             NaN              NaN
2702512        14543  Rock Springs  ...             NaN              NaN

[2702513 rows x 11 columns]

ตรวจสอบ missing


## ทำการลบ missing โดยใช้ **Dropna**

In [66]:
merged_table.isnull().any()

รหัสสนามบิน           False
เมือง                 False
รัฐ                   False
ชื่อสนามบิน           False
วันในเดือน             True
วันในสัปดาห์           True
สายการบิน              True
รหัสสนามบินต้นทาง      True
รหัสสนามบินปลายทาง     True
ความล่าช้าขาออก        True
ความล่าช้าขาเข้า       True
dtype: bool

In [76]:
merged_table_dropna = merged_table.dropna()
merged_table_dropna

วันในเดือน  ...                                 ชื่อสนามบิน
0                19  ...                  Detroit Metro Wayne County
1                19  ...                Salt Lake City International
2                19  ...                      Portland International
3                19  ...             Lambert-St. Louis International
4                19  ...  Cincinnati/Northern Kentucky International
...             ...  ...                                         ...
2702213          19  ...                       Orlando International
2702214          19  ...    Hartsfield-Jackson Atlanta International
2702215          19  ...           Albuquerque International Sunport
2702216          19  ...    Hartsfield-Jackson Atlanta International
2702217          19  ...    Hartsfield-Jackson Atlanta International

[2702218 rows x 11 columns]

In [77]:
merged_table_dropna.shape #จะเห็นได้ว่าหลังจากdropna มีข้อมูลของเเถวหายไปเป็นเเถวที่มีช่องว่าง Nan


(2702218, 11)

ขาเข้ามากสุด เเละ ขาเข้าน้อยสุด

In [78]:
total = merged_table_dropna.groupby('รัฐ')['ความล่าช้าขาเข้า'].max()
total = pd.DataFrame(total)
total.reset_index()  #นำตารางที่เป็น dropna มาเลือกเเค่คอลัมที่เราต้องการ


รัฐ  ความล่าช้าขาเข้า
0   AK               612
1   AZ              1835
2   CA              1385
3   CO              1143
4   CT              1227
5   DC              1137
6   FL              1466
7   GA              1429
8   HI              1228
9   IL              1081
10  IN              1423
11  KY              1064
12  LA               894
13  MA              1146
14  MD              1175
15  MI              1179
16  MN              1436
17  MO               988
18  NC              1440
19  NE              1301
20  NJ               875
21  NM              1391
22  NV              1845
23  NY              1127
24  OH              1122
25  OK              1134
26  OR              1330
27  PA              1475
28  PR               842
29  RI               631
30  TN              1351
31  TX              1233
32  UT              1609
33  VA              1114
34  WA              1421
35  WI              1032

In [79]:
max(total['ความล่าช้าขาเข้า'])

1845

In [80]:
total.describe()

ความล่าช้าขาเข้า
count         36.000000
mean        1225.583333
std          274.329664
min          612.000000
25%         1105.750000
50%         1203.000000
75%         1421.500000
max         1845.000000

In [81]:
total[total['ความล่าช้าขาเข้า'] == 1845]  # ดั้งนั้นรัฐที่มีจำนวนเที่ยวบินขาเข้ามากที่สุดคือ รัฐ NV

ความล่าช้าขาเข้า
รัฐ                  
NV               1845

In [82]:
total_min = merged_table_dropna.groupby('รัฐ')['ความล่าช้าขาเข้า'].min()
total_min = pd.DataFrame(total_min)
total_min.reset_index()  #นำตารางที่เป็น dropna มาเลือกเเค่คอลัมที่เราต้องการ

รัฐ  ความล่าช้าขาเข้า
0   AK               -59
1   AZ               -54
2   CA               -67
3   CO               -58
4   CT               -61
5   DC               -71
6   FL               -57
7   GA               -63
8   HI               -62
9   IL               -66
10  IN               -49
11  KY               -45
12  LA               -51
13  MA               -71
14  MD               -60
15  MI               -56
16  MN               -66
17  MO               -54
18  NC               -69
19  NE               -41
20  NJ               -86
21  NM               -51
22  NV               -67
23  NY               -79
24  OH               -58
25  OK               -49
26  OR               -53
27  PA               -65
28  PR               -51
29  RI               -94
30  TN               -50
31  TX               -58
32  UT               -61
33  VA               -51
34  WA               -62
35  WI               -65

In [83]:
total_min.describe()

ความล่าช้าขาเข้า
count         36.000000
mean         -60.555556
std           10.895463
min          -94.000000
25%          -66.000000
50%          -59.500000
75%          -52.500000
max          -41.000000

In [84]:
min(total_min['ความล่าช้าขาเข้า'])

-94

In [86]:
total_min[total_min['ความล่าช้าขาเข้า'] == -94.0]

ความล่าช้าขาเข้า
รัฐ                  
RI                -94

ตอบจำนวนเครื่องบิน ขาเข้ามากที่สุดคือรัฐ NV 1845 และรัฐที่มีเที่ยวบินขาเข้าน้อยที่สุดคือรัฐ RI -94

## **ข้อ 3 สายการบินไหนมีเที่ยวบินมากที่สุดและน้อยที่สุด**

In [74]:
merged_table = flights.merge(airports,how='left',left_on='รหัสสนามบินต้นทาง',right_on='รหัสสนามบิน')
merged_table


วันในเดือน  ...                                 ชื่อสนามบิน
0                19  ...                  Detroit Metro Wayne County
1                19  ...                Salt Lake City International
2                19  ...                      Portland International
3                19  ...             Lambert-St. Louis International
4                19  ...  Cincinnati/Northern Kentucky International
...             ...  ...                                         ...
2702213          19  ...                       Orlando International
2702214          19  ...    Hartsfield-Jackson Atlanta International
2702215          19  ...           Albuquerque International Sunport
2702216          19  ...    Hartsfield-Jackson Atlanta International
2702217          19  ...    Hartsfield-Jackson Atlanta International

[2702218 rows x 11 columns]

## จากตาราง merged_table จะหาสายการบินมากที่สุด 


In [87]:
merged_table.groupby('สายการบิน').count()

วันในเดือน  วันในสัปดาห์  ...     รัฐ  ชื่อสนามบิน
สายการบิน                            ...                     
9E              80031         80031  ...   80031        80031
AA             289855        289855  ...  289855       289855
AS              68555         68555  ...   68555        68555
B6             121906        121906  ...  121906       121906
DL             381657        381657  ...  381657       381657
EV             157928        157928  ...  157928       157928
F9              35738         35738  ...   35738        35738
FL              92702         92702  ...   92702        92702
HA              17432         17432  ...   17432        17432
MQ             113212        113212  ...  113212       113212
OO             160164        160164  ...  160164       160164
UA             286418        286418  ...  286418       286418
US             233321        233321  ...  233321       233321
VX              34739         34739  ...   34739        34739
WN             575739        575739  ...  575739       575739
YV              52821         52821  ...   52821        52821

[16 rows x 10 columns]

ให้ค่า sort values เป็น วันในเดือน โดยให้ค่า ascending เป็น False เพื่อให้เรียงจากมากลงไปน้อย

In [88]:
summary = merged_table.groupby('สายการบิน').count().sort_values(by=['วันในเดือน'],ascending=False)
summary

วันในเดือน  วันในสัปดาห์  ...     รัฐ  ชื่อสนามบิน
สายการบิน                            ...                     
WN             575739        575739  ...  575739       575739
DL             381657        381657  ...  381657       381657
AA             289855        289855  ...  289855       289855
UA             286418        286418  ...  286418       286418
US             233321        233321  ...  233321       233321
OO             160164        160164  ...  160164       160164
EV             157928        157928  ...  157928       157928
B6             121906        121906  ...  121906       121906
MQ             113212        113212  ...  113212       113212
FL              92702         92702  ...   92702        92702
9E              80031         80031  ...   80031        80031
AS              68555         68555  ...   68555        68555
YV              52821         52821  ...   52821        52821
F9              35738         35738  ...   35738        35738
VX              34739         34739  ...   34739        34739
HA              17432         17432  ...   17432        17432

[16 rows x 10 columns]

จากผลลัพธ์ที่ได้นำมาหาค่า Max และ Min เพื่อหาเที่ยวบินที่มากสุดแะเที่ยวบินน้อยสุดจะได้

In [89]:
max = summary.iloc[:1,:1]
max

วันในเดือน
สายการบิน            
WN             575739

พบว่าสายการบิน WN มีเที่ยวบินที่มากที่สุด =575739 ครั้ง


In [90]:
min = summary.iloc[15:16,:1]
min

วันในเดือน
สายการบิน            
HA              17432

พบว่าสายการบิน WN มีเที่ยวบินที่น้อยที่สุด =17432 ครั้ง

## **สร้างโจทย์ใหม่1ข้อ**

สายการบิน ใด บินไปยังเมือง Salt Lake City International มากที่สุด

In [92]:
data = merged_table[merged_table['รหัสสนามบิน']==14869]
data

วันในเดือน  วันในสัปดาห์  ... รัฐ                   ชื่อสนามบิน
1                19             5  ...  UT  Salt Lake City International
27               19             5  ...  UT  Salt Lake City International
32               20             6  ...  UT  Salt Lake City International
34               20             6  ...  UT  Salt Lake City International
56               20             6  ...  UT  Salt Lake City International
...             ...           ...  ...  ..                           ...
2702156          19             6  ...  UT  Salt Lake City International
2702160          19             6  ...  UT  Salt Lake City International
2702175          19             6  ...  UT  Salt Lake City International
2702188          19             6  ...  UT  Salt Lake City International
2702208          19             6  ...  UT  Salt Lake City International

[47324 rows x 11 columns]

In [96]:
Max_3 = data.sort_values(by=['เมือง'],ascending=False)
Max_3.iloc[:2,[2,8]]

สายการบิน           เมือง
1              DL  Salt Lake City
1821808        DL  Salt Lake City

ตอบ สายการบินที่บินไปยังเมือง Salt Lake City มากที่สุดคือสายการบินของ DL